peps/pep-0735.rst

```diff
@@ -0,0 +1,490 @@
+    typing = ["mypy", "types-requests", ".[types]"]
+    typing-test = ["[typing,test]", "useful-types"]
+
+    [project.optional-dependencies]
+    types = ["typing-extensions"]
+
+Note how ``test`` and ``typing`` are able to refer to the current package
+while ``docs`` does not. This reflects the ability of Dependency Groups to be used
+in the same manner as extras, adding to dependencies, or completely
+independently. ``typing-test`` is defined as a union of two existing groups,
+plus an additional package. ``typing`` includes an extra, ``types``.
+
+Implementation Requirements
+---------------------------
+
+Build backends MUST NOT include dependency group data in built distributions.
+
+When installing the current package (``.``) from a dependency group, tools
+SHOULD prefer editable installs over non-editable installs. They MAY provide
+users with options to configure this behavior.
+
+Use of Dependency Groups
+------------------------
+
+Tools which support Dependency Groups are expected to provide new options and
+interfaces to allow users to install from Dependency Groups. Implementations
+may wish to treat them similarly to their current treatments of
+``requirements.txt`` files, or more similarly to extras.
+
+No syntax is defined for expressing the dependency group of a package, for two
+reasons:
+
+* it would not be valid to refer to the dependency groups of a third-party
+  package from pypi (because the data is defined to be unpublished)
+
+* there is not guaranteed to be a current package for dependency groups -- part
+  of their purpose is to support non-package projects
+
+For example, a possible pip interface for installing dependency groups
+would be:
+
+.. code:: shell
+
+    pip install --dependency-groups=test,typing
+
+Note that this is only an example. This PEP does not declare any requirements
+for how tools support the installation of Dependency Groups.
+
+Reference Implementation
+========================
+
+There is currently no reference implementation/consumer of this specification.
+
+Backwards Compatibility
+=======================
+
+At time of writing, the ``requirements`` namespace within a
+``pyproject.toml`` file is unused. Since the top-level namespace is
+reserved for use only by standards specified at packaging.python.org,
+there should be no direct backwards compatibility concerns.
+
+Security Implications
+=====================
+
+This PEP introduces new syntaxes and data formats for specifying dependency
+information in projects. However, it does not introduce newly specified
+mechanisms for handling or resolving dependencies.
+
+It therefore does not carry security concerns other than those inherent in any
+tools which may already be used to install dependencies -- i.e. malicious
+dependencies may be specified here, just as they may be specified in
+``requirements.txt`` files.
+
+How to Teach This
+=================
+
+This feature should be referred to by its canonical name, "Dependency Groups".
```

Comment:
... which is very different from the entry you see in pyproject.toml, `requirements.packages`. Maybe one of the two should change.

Reply:
I'm going to be changing the table name to `[dependency-groups]` today. Which will force me to do some rewrites throughout.

I wanted to use `[project.dependency-groups]` but Paul very rightly pointed out that I then need to reckon with the fact that `project.name` and `project.version` are required when `project` is present. Even though I still like that name best, I think it's biting off more than I can chew.